### DISCUSSION TITLE
Need more commitment here.

### DISCUSSION THREAD

**Discussion by najak3d** [dotnet/Microsoft.Maui.Graphics.Controls#103] [General]
Hi @jsuarezruiz ,  in order to gauge whether or not we should put our eggs into the "Maui.Graphics.Controls" basket, we need to know more about Microsoft's planned allocation of labor for this project (i.e. "FTE" = full-time-equivalents). 

I'm not sure if you only are given 10 hrs/week to work on this, or 40 hrs..  And if there are plans to add to the "Graphics.Controls" team.  We need to know the truth in this matter -- how serious is Microsoft about producing a "Graphics.Controls" library which is "production ready" -- and how soon do they estimate before it's in Beta?

If instead Microsoft is planning for it to be implemented mostly by volunteers in Open-Source environment, then we need to know that as well, and then need to figure if this is a feasible plan.

I love, love, love the idea of Microsoft producing a "pixel-perfect-solution" for all to use.   IMO, it's their most important Add-on to MAUI, in that it will apply across-the-board for all platforms/scenarios --  iOS, Android, Windows, Mac, Linux, and Web browsers -- AND... best of all -- GAMES!

It was a breeze to integrate Avalonia with Urho.NET.   Similarly, Maui.Graphics.Controls offers this capability as well, as we can simply render the output to a GPU texture that is rendered in the Game UI.    This can be done not only for Urho.NET, but also for ANY Game Engine that integrates to .NET/Mono.

I've been trying to put the breaks on our effort with Avalonia, to determine if Maui.Graphics.Controls will be ready in time, and we do NOT have enough resources to spend months *writing the missing parts of Graphics.Controls!*    So, we need to know how much effort/money Microsoft is going to pour into this, so that we can set our expectations more accurately, and make decisions based on the truth, rather than fantasy.

In short, can you please tell us Microsoft's commitment to Graphics.Controls (in terms of "FTE", investment, and intentions for how this work will get done, and when).   Even ballpark estimates/guesses would be great.  Thanks!

**Comment by jsuarezruiz** [comment]
As you know, at the moment it is an experimental library and the main goal right now is to validate interest, etc. Really appreciate your interest and feedback but Let me include here @davidortinau and @Redth  who can maybe share more details.

**Reply by najak3d** [reply]
What are the requirements for "clearing the bar" to get this better funded?

Looking at the YouTube views on various videos -- looks like we've got 3000 views on your Graphics.Controls libraries.   Part of this is lack of advertising, IMO.

The other indicators that Pixel-Perfect solutions are highly desired come from the dramatic uptake in devs switching over to ***Flutter/Dart*** which mainly offers a Pixel-Perfect solution.    Also, the popularity/momentum of ***Avalonia***, also indicates that Pixel-Perfect is widely accepted/desired.

In the beginning, iOS used to DEMAND that apps maintain THEIR UI-style, in order to be approved on the App Store.   But they saw-the-light, and have been accepting Pixel-Perfect solutions since 2012, and have gotten more lenient over the years.

IIRC, the main reason Xamarin-Forms (and now MAUI) mainly enforced usage of Native controls, was to satisfy this desire for apps to look and behave differently on each platform. 

Our app, which has been Pixel-Perfect, has been running on WinCe, Windows, iOS, and Android since 2012, and in both the AppStore and PlayStore.     It is awesome only having ONE VERSION of User Guide, and having the same exact look/behavior on all platforms.

This should become the **_primary_** way that future apps will be developed on MAUI -- Pixel-Perfect, because it simply makes the most sense, for most apps, IMO.

**Reply by najak3d** [reply]
How would you like for me to send you my code for inclusion on Github?  (e.g. pull-request?)

For starters, I'd like to send you my code rewrites of DrawMapper, GraphicsControlHander, and the GalleryPageCSharp.

Next will be MixedGraphicsControlHandler.

**Reply by jsuarezruiz** [reply]
David or Jonathan are the right people to be able to answer this. For my part, I keep the library up to date, try to clear doubts and collect feedback, so thank you for sharing your feedback.

About how to send code, definitely PRs!. Is a good way to open discussions and improve.

**Reply by najak3d** [reply]
> David or Jonathan are the right people to be able to answer this. For my part, I keep the library up to date, try to clear doubts and collect feedback, so thank you for sharing your feedback.

Can you get either of them to answer here?   I'm hearing a lot of bad-mouthing MAUI elsewhere, and am arguing against them.  I want to know how devoted Microsoft is to MAUI success.   For the decade prior to MAUI, Microsoft's offering to for .NET has faltered, compared to the years when Bill Gates was still guiding the company.    Much of the dev community has grown (more?) cynical about Microsoft's willingness to actually deliver something that competes or beats the competition (like Flutter/DART). 

I'm holding out hope and optimism that MAUI marks Microsoft's return to the competition.  I've always been a Microsoft fan, which started mostly with the amazing progress of .NET around 2003 time frame, which continued strong until about 2010 -- when things started to falter (regarding .NET).   This opened the way for competitors to take market share; I could be wrong, but this seems to be a fairly common view.

If MAUI turns out great -- all that becomes water-under-the-bridge.   And Graphics.Controls, needs to succeed as well.

In the absence of Graphics.Controls being usable, we're turning to Avalonia instead, which is a powerhouse open-source competitor to MAUI.   But if MAUI delivers -- this is where we want to put our eggs.

**Reply by najak3d** [reply]
One great thing about Avalonia is that they've made it easy for us to insert our own Renderer -- which allowed us to make Urho.NET into the Renderer, so that we can now render Avalonia UI inside of the Urho.NET Game window.  Impressive stuff. 

But for me, as we find this success, I'm still just hoping that MAUI will step up, and give us a good reason to switch over to MAUI instead.   This integration with Avalonia wasn't too hard, but so far, on the MAUI forums, I've posted a question about this over a month ago -- and NO ONE responded.  No answers.  That's not encouraging.

**Reply by najak3d** [reply]
In short, I don't plan on spending much time here, or trying to contribute, unless I see MAUI become stable and awesome -- as they are set up to do, and also see Microsoft step up the creation of Graphics.Controls.  I believe one talent FTE dedicated to Graphics.Controls for a year or two is all it would take to make this a winning solution.

**Reply by davidortinau** [reply]
Hi @najak3d, I think Avalonia is the right choice for you given what I'm reading here. Our top priority is to ship .NET MAUI itself at a high level of quality to serve our existing customers. It's too early yet in the experiment to give you the confidence and details you're asking for. I agree with you about the importance and value of this project. I hope you'll keep an eye on Graphics.Controls as it evolves, and share your thoughts as you've done here.

**Reply by najak3d** [reply]
David, thank you for responding here. 

I will definitely check back periodically, and will try to keep our options open to moving to MAUI in the future.   Avalonia's XAML (and Code Markup) language is quite close to that of MAUI, so switching UI's shouldn't be too complex, if it is warranted.

I wish Microsoft would stop with the "experimenting" and just invest more money into MAUI, including Comet and Graphics.Controls.  It seems to me a no-brainer that these are winning solutions.    The more Microsoft drags their feet to "experiment", the worse the situation gets (as MAUI loses market share to platforms like Flutter, which support MVU, and promote primarily a Pixel-Perfect UI solution).

Delaying Graphics.Controls is like shooting yourself in the foot.  There is money to be saved by any multi-platform project that could write-once and actually have the UI look/behave the EXACT SAME on all platforms.   It's a beautiful concept that should be PRIORITIZED by Microsoft now, not later.   I'm guessing many on your team feel the same way, but it comes down to the purse strings and budget.

I think someone should light a small fire under those who control this budget.

**Reply by saint4eva** [reply]
You are still repeating yourself.

**Reply by najak3d** [reply]
> You are still repeating yourself.

Yes, I repeat because I think these points need to be **shouted** from all the rooftops and heard by Microsoft.  As more time passes, the potential gains decay, as more projects switch over to Flutter/Dart and Avalonia, for their **Pixel-Perfect** solutions.  Pixel-Perfect is far easier-to-achieve than Maui's current core focus of wrapping native controls on each platform.

It seems that having just one FTE working on Graphics.Controls would do wonders for the timeliness and final outcome for Maui as a whole.   

Avalonia is matured open-source, and in large part, could be mimicked for it's architecture/design, but it would take a dedicated FTE developer to make this a reality, in a timely fashion.

Our team is switching to "Avalonia + Stride3D" for our Pixel-Perfect solution that renders Avalonia to the Game Window UI.

**Reply by najak3d** [reply]
Since it always comes down to budgets, I think Microsoft should consider using "Open Collective" site as a means for allowing businesses/teams/people to contribute $$$ to pay towards the development of things they find most important.   I wouldn't be surprised to see $100K raised for Graphics.Controls, if it were advertised.

I would like to see Microsoft do "100%" match for these donations, so $100K doubles to be $200K, and buys the community 2 FTE's for a year.

https://opencollective.com/

**Reply by saint4eva** [reply]
> > You are still repeating yourself.
> 
> Yes, I repeat because I think these points need to be **shouted** from all the rooftops and heard by Microsoft. As more time passes, the potential gains decay, as more projects switch over to Flutter/Dart and Avalonia, for their **Pixel-Perfect** solutions. Pixel-Perfect is far easier-to-achieve than Maui's current core focus of wrapping native controls on each platform.
> 
> It seems that having just one FTE working on Graphics.Controls would do wonders for the timeliness and final outcome for Maui as a whole.
> 
> Avalonia is matured open-source, and in large part, could be mimicked for it's architecture/design, but it would take a dedicated FTE developer to make this a reality, in a timely fashion.
> 
> Our team is switching to "Avalonia + Stride3D" for our Pixel-Perfect solution that renders Avalonia to the Game Window UI.

Why not use the so-called project/ platform that works well for you?

**Reply by saint4eva** [reply]
> Since it always comes down to budgets, I think Microsoft should consider using "Open Collective" site as a means for allowing businesses/teams/people to contribute $$$ to pay towards the development of things they find most important. I wouldn't be surprised to see $100K raised for Graphics.Controls, if it were advertised.
> 
> I would like to see Microsoft do "100%" match for these donations, so $100K doubles to be $200K, and buys the community 2 FTE's for a year.
> 
> https://opencollective.com/

Nothing stops you from contributing to those projects. PRs are accepted.

**Reply by najak3d** [reply]
I have more money to contribute (as business expense) than I have time to spend contributing to MAUI.

And I'm concerned about the long-term outlook.  I need something sooner, not later.  Therefore, our team (along with many others) are forced to go elsewhere to obtain a reasonable "Pixel Perfect" UI solution (i.e Avalonia for us, and Flutter/Dart for many others).   I would have much preferred to use MAUI, given that it's backed by Msft, and not just open-source good will contributors.    A platform backed by paid-developers is a superior paradigm, usually.   And so, it's Avalonia for us, and we are lucky that the devs there are working full-time and very talented and dedicated.

And so I'm screaming here - because I think Msft is losing market share, due to their failure to put more effort/commitment into their own Pixel Perfect solution -- i.e. Maui.Graphics.Controls.    It's a shame.

**Reply by saint4eva** [reply]
Good luck with you and your team. If any platform and tool set serve your organisation's need, kindly use it. No need to wait for Maui.

**Reply by alaingaudreau** [reply]
@jsuarezruiz @davidortinau  @Redth  We are now four months from this thread's start and we are pretty much at the same point and Windows- isn't even available with the nightly's making it impossible for us to test on our principal build target.

We build custom business software for our clients and have a few products in the works which will be public commercial products that we have been hoping to be able to use .net and maui to provide multi platform on a single codebase but it's proving to be impossible.

I get that maui and maui graphic controls are still pre-release (1 year? 2 years past due for maui?) but at some point if you want us to adopt new technology for any concrete development other than toy mobile apps, there needs to be more movement.

Right now, MS is giving us the impression that maui and maui graphics controls will be replaced before even being released as it's taking so long.

It's crippling enough for business apps that you are not planing on having a DataGridView compatible control on the main branch, without easy access to create our own by reusing some of the supporting code found in this project, it is a major undertaking to build one from scratch with all the required features to be useable in a commercial product.

Building a quality control like a datagrid is a project all on it's own with months of development to get it just right...

From all the replies you have been making here and on other threads, you keep repeating you are validating interest which is great but how exactly are we supposed to get a feel for it when we can't even use it?

Right now you have thousands of developers waiting for movement on your part to be able to assess if maui is even a viable project or if the risk that it never makes it to release is too high to justify the learning curve.

We in the commercial world, need to see more concrete actions, reliable timelines and working libraries to test with.

It's getting hard to take maui seriously and by consequence, microsoft development technology as well, especially when like right now, this project does not even work on the company's primary and the only supported platform it owns.

**Reply by saint4eva** [reply]
There are many commercial control vendors that would provide you with all you need.

**Reply by alaingaudreau** [reply]
@saint4eva syncfusion, telerik and devexpress are all using these graphics libs for their grids which means they are waiting for the team to step up as well. 

devexpress has a datagrid out now but supports mobile only at the moment.

hypothetically this or that company may release datagrids and other controls for all the platforms supported by maui but if there is no or little commitment by microsoft to these essential base projects, we will end up with just another toolkit that is replaced before it's acceptance like others that came before it.

we can't base business decisions on perhaps and maybe's.

these are decisions that will affect businesses for years to come, take up lots of time and will require hundreds of thousands of dollars in investments for training, tooling, testing etc

we need for microsoft to step up to the plate and show us maui will have the funding and human resources to back maui and these base projects for us to assess it's viability for the commercial use of this framework.

right now, it's floundering like so many open source projects that end up being abandoned after 6 months or a year...

we need more predictability and transparency from microsoft.

**Reply by davidortinau** [reply]
@alaingaudreau I appreciate your passion for this experiment! This is why we released the code, to receive this feedback. 

It would be very helpful to get some specific details from you about how drawn UI controls would meet your app business objectives. Let's discuss david.ortinau@microsoft.com

> we need more predictability and transparency from microsoft.

I want to make sure you are seeing what we are publishing. We work hard to be transparent and visible.

https://themesof.net - .NET public roadmaps
https://dotnet.microsoft.com/en-us/platform/support/policy - .NET support policy and release schedule

.NET MAUI is officially supported since we shipped RC1 in April this year. GA will at Microsoft Build 2022. From here on .NET MAUI is on the .NET release schedule.

This repository (Microsoft.Maui.Graphics.Controls) is an experiment as stated on the main page. It is unsupported.

You mention the need for libraries, and this is the primary call to action now. We have heard from many library maintainers that they have been waiting for RC and GA of .NET MAUI to work on transitioning their libs. That includes expanding some from mobile to desktop. In just the past week we've seen many new releases of libraries with .NET MAUI support.

Re: controls like DataGrid, we haven't put any new controls on our roadmap. The focus of the product is quality, stability, and supporting the ecosystem in this initial transition to .NET 6. We will be considering more controls in the future, and in the meantime look to control vendors to meet needs. Note, you can always use any platform native controls in .NET MAUI by writing your own handler (to use the WinUI DataGrid for example).

**Comment by tcsaddul** [comment]
I am also waiting for non-experimental Comet for our next generation apps.

**Reply by saint4eva** [reply]
Clancey and some other guys are working hard towards achieving that. And moreover, some apps built with Comet are already published in Play store and App store.

**Reply by najak3d** [reply]
It seems to me that Comet was far more matured/complete than Graphics.Controls.   Both are vital to MAUI success.

**Comment by BC89** [comment]
At what point can we expect Microsoft to make a commitment to this effort and move it beyond experimental or kill it? I would hope one day soon this becomes the default engine for Maui. From the developers and early-adopters I've spoken to, the consensus is that out of the box Maui UX are way too poor for a proper product in today's market. End users expect a lot more spit and polish than Maui can offer and it's one reason Flutter is killing it. Being a life long C# engineer I know there's more that goes in to selecting a product platform but sadly the UX is a real blocker.

**Reply by manishbisen** [reply]
Time come for Microsoft to move experiment projects to beyond steps.Hope they will listen it.

**Reply by saint4eva** [reply]
After .NET Maui team has shipped Maui, they will start focusing on other areas such as drawn controls and MVU C# (Comet)

**Comment by BC89** [comment]
Take a long look at flutter. We are an MS shop but Maui at this point (however late) is embarrassingly lacking in the presentation department. My sense is the lack of commitment may be a tacit acknowledgement by MS bean counters that google ate their lunch.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: alaingaudreau ***@***.***>
Sent: Saturday, May 21, 2022 9:03:47 AM
To: dotnet/Microsoft.Maui.Graphics.Controls ***@***.***>
Cc: Brendan Carroll ***@***.***>; Comment ***@***.***>
Subject: [External Email] Re: [dotnet/Microsoft.Maui.Graphics.Controls] Need more commitment here. (Discussion #103)


<https://github.com/saint4eva>

***From ***@***.*** 410-316-7820 *** This is an External Email from outside of KCI.***
________________________________
@saint4eva syncfusion, telerik and devexpress are all using these graphics libs for their grids which means they are waiting for the team to step up as well.

devexpress has a datagrid out now but supports mobile only at the moment.

hypothetically this or that company may release datagrids and other controls for all the platforms supported by maui but if there is no or little commitment by microsoft to these essential base projects, we will end up with just another toolkit that is replaced before it's acceptance like others that came before it.

we can't base business decisions on perhaps and maybe's.

these are decisions that will affect businesses for years to come, take up lots of time and will require hundreds of thousands of dollars in investments for training, tooling, testing etc

we need for microsoft to step up to the plate and show us maui will have the funding and human resources to back maui and these base projects for us to assess it's viability for the commercial use of this framework.

right now, it's floundering like so many open source projects that end up being abandoned after 6 months or a year...

we need more predictability and transparency from microsoft.

**Reply by davidortinau** [reply]
If we were to fund and incorporate this controls project into .NET MAUI, how would that change things for you?

Can you give me some examples of what you mean by presentation?

**Reply by davidortinau** [reply]
And "spit and polish", and "ux is a real blocker"?

**Reply by najak3d** [reply]
> If we were to fund and incorporate this controls project into .NET MAUI, how would that change things for you?
> 
> Can you give me some examples of what you mean by presentation?

It would have significantly changed the existing state of our application.  Right now, we're only released to Alpha test for Android, because in iOS "many things aren't working" (UI bugs of various sorts, including layout or behavior).  All because Xamarin Forms is wrapping iOS native controls and has nuances (that we haven't resolved yet, but will soon).

In the last year, we ported our 2012 app that used our own self-built pixel-perfect UI, to Xamarin Forms as part of a facelift, and updating to the latest tech.   It was a gross disappointment and a shocking experience (compared to the quality of Microsoft stuff that we adopted pre-2010).

At this point, this current project of ours will probably not use Pixel-Perfect; that ship has sailed.  But had it been available a year ago, we would have most certainly used it.

If Pixel-Perfect for MAUI gains commitment/funding from Microsoft, then we would put our eggs into this basket for future projects, which include mobile games.   Since it's easy to make SkiaSharp directly modify OpenGL Texture2D objects on the GPU, this enables an efficient pipeline between Maui.Graphics.Controls into the gaming world, with just a few thousand lines of code to create the adapter layer (that's how we did it for Urho.NET and Avalonia).

**Reply by charlesroddie** [reply]
@davidortinau Currently we use XF with a small list of collections/controls to reduce the bug surface area we are exposed to.

To reduce bugs and platform inconsistencies further (and open up the web in future) we have implemented 1/3 of our UI so far in SkiaSharp and are migrating the rest. It would have saved a huge amount of effort had there been drawn controls at the start in XF, in terms of reduced bugs and inconsistencies.

Drawn controls gives people choice, and most developers would choose the combination of pixel-perfect views/few bugs/high perf over native views/many bugs/low perf.

Since it is would be a fundamental choice in MAUI I don't think there would be a problem getting this information out and it would reduce the reputation of MAUI as a buggy platform.

I am estimating the time before MAUI becomes less buggy than current XF, at which point we could adopt it. I am thinking 1 year based on feedback from other devs. This is very long for us given that XF is stuck on mono (which we get platform-specific bugs from that would probably be fixed in dotnet7).

It seems counterintuitive that adding an option could stabilize something, but for drawn controls I believe this could be the case, and you could get what we would consider a usable version of MAUI ready in perhaps 6 months.

**Reply by jsuarezruiz** [reply]
> @davidortinau Currently we use XF with a small list of collections/controls to reduce the bug surface area we are exposed to.
> 
> To reduce bugs and platform inconsistencies further (and open up the web in future) we have implemented 1/3 of our UI so far in SkiaSharp and are migrating the rest. It would have saved a huge amount of effort had there been drawn controls at the start in XF, in terms of reduced bugs and inconsistencies.
> 
> Drawn controls gives people choice, and most developers would choose the combination of pixel-perfect views/few bugs/high perf over native views/many bugs/low perf.
> 
> Since it is would be a fundamental choice in MAUI I don't think there would be a problem getting this information out and it would reduce the reputation of MAUI as a buggy platform.
> 
> I am estimating the time before MAUI becomes less buggy than current XF, at which point we could adopt it. I am thinking 1 year based on feedback from other devs. This is very long for us given that XF is stuck on mono (which we get platform-specific bugs from that would probably be fixed in dotnet7).
> 
> It seems counterintuitive that adding an option could stabilize something, but for drawn controls I believe this could be the case, and you could get what we would consider a usable version of MAUI ready in perhaps 6 months.

It would be interesting to know what Skia controls you have created, the needs covered, the ones you have, etc. You can write me an email if you want to share more details :)

**Comment by tcsaddul** [comment]
It is a pity these projects like Comet are just "personal" experimental.

**Comment by RChrisCoble** [comment]
@davidortinau there are other folks like myself that are looking to use pixel perfect controls in Blazor WASM as well, along with future support for Blazor Server (and Hybrid).  MAUI.Graphics and MAUI.Graphics.Controls could be a huge deal that benefits not just MAUI, but the larger UX ecosystem in general.
